dna/zomes/courses/code/src/course.rs

```diff
@@ -21,119 +21,223 @@ use std::convert::TryFrom;
+                },
+                validation: | _validation_data: hdk::LinkValidationData | {
+                    // TODO:  should validate that we're only linking against CourseData that has
+                    // this entry's address in course_anchor field
+                    Ok(())
+                }
+            ),
+            from!( // to query all the courses of a user(all courses that a user is the teacher or owner of)
+                "%agent_id",
+                link_type: "teacher->courses",
+                validation_package: || {
+                    hdk::ValidationPackageDefinition::Entry
+                },
+                validation: | _validation_data: hdk::LinkValidationData | {
+                    // TODO: Homework. Implement validation rules if required.
+                    Ok(())
+                }
+            ),
+            from!( // to query all courses that one user enrolled
+                "%agent_id",
+                link_type: "student->courses",
+                validation_package: || {
+                    hdk::ValidationPackageDefinition::Entry
+                },
+                validation: | _validation_data: hdk::LinkValidationData | {
+                    // TODO: Homework. Implement validation rules if required.
+                    Ok(())
+                }
+            ),
+            to!( // to query all enrolled user for a course
+                "%agent_id",
+                link_type: "course->students",
+                validation_package: || {
+                    hdk::ValidationPackageDefinition::Entry
+                },
+                validation: | _validation_data: hdk::LinkValidationData | {
+                    // TODO: Homework. Implement validation rules if required.
+                    Ok(())
+                }
+            )
       ]
     )
 }
 
-//// Anchor Definition : This Anchor will be used to query all courses
-pub fn anchor_entry_def() -> ValidatingEntryType {
+/* *************** CourseData HDK entry definition */
+pub fn course_data_def() -> ValidatingEntryType {
     entry!(
-        name: "anchor",
+        name: "course_data",
+        description: "Entry type that holds all course data that could be modified",
+        sharing: Sharing::Public,
+        validation_package: || {
+            hdk::ValidationPackageDefinition::Entry
+        },
+        validation: | validation_data: hdk::EntryValidationData<CourseData>| {
+            match validation_data{
+                EntryValidationData::Create { entry, validation_data } => {
+                    validate_only_teacher_can_do(validation_data.sources(), &entry.teacher_address, "create")?;
+                    validate_course_title(&entry.title)
+                },
+                EntryValidationData::Modify { new_entry, old_entry, validation_data, .. } => {
+                    validate_only_teacher_can_do(validation_data.sources(), &old_entry.teacher_address, "modify")?;
```

Comment:
I'd remove the teacher_address from course_data, and here in the validation I'd query the course_anchor to get the teacher_address and see if they are signing the course_data creation/modification/deletion. Here it is safe to do get_entry on the anchor since it will never change.

Reply:
@guillemcordoba  I don't think removing the teacher_address from the course_data would bring any benefits, and querying another entry on every validation would worsen the performance (however not that much, I suppose).
Why do you think we should do that? I'm not sure I understand your reasons